src/main/java/de/uniba/dsg/serverless/provider/google/GoogleLogHandler.java

```diff
@@ -0,0 +1,185 @@
+	 * therefore a self-assigned key within the cloud function in form of a uuid is
+	 * used to achieve the mapping from platform executions and local REST calls.
+	 */
+	@Override
+	public Map<String, WritableEvent> getPerformanceData() throws SeMoDeException {
+
+		List<LogEntry> logEntries = this.executeListLogEntries();
+
+		Map<String, List<LogEntry>> eventMap = this.generateEventMapByExecutionId(logEntries);
+
+		return this.generatePerformanceDataMap(eventMap);
+	}
+
+	private Map<String, WritableEvent> generatePerformanceDataMap(Map<String, List<LogEntry>> eventMap) {
+
+		Map<String, WritableEvent> performanceMap = new HashMap<>();
+		
+		for(String key : eventMap.keySet()) {
+			List<LogEntry> cohesiveEvent = eventMap.get(key);
+			PerformanceData data = this.extractPerformanceData(cohesiveEvent);
+			performanceMap.put(data.getRequestId(), data);
+		}
+		
+		return performanceMap;
+	}
+
+	// TODO: make the whole function more robust -> refactor
+	private PerformanceData extractPerformanceData(List<LogEntry> cohesiveEvent) {
+		
+		LocalDateTime startTime = LocalDateTime.MIN;
+		LocalDateTime endTime = LocalDateTime.MIN;
+		double preciseDuration = -1.0;
+		String platformId= "";
+		String instanceId = "";
+		
+		for(LogEntry entry : cohesiveEvent) {
+			Payload payload = entry.getPayload();
+			if(payload instanceof StringPayload) {
+				String data = ((StringPayload)payload).getData();
+				System.out.println(data);
+				if(data.startsWith("Function execution took")) {
+					preciseDuration = Double.parseDouble(data.split(" ")[3]);
+					endTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(entry.getTimestamp()), ZoneId.systemDefault());
+				}else if(data.startsWith("Function execution start")) {
+					startTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(entry.getTimestamp()), ZoneId.systemDefault());
+				}else if(data.startsWith("SEMODE::")) {
+					ObjectMapper om = new ObjectMapper();
+					try {
+						JsonNode node = om.readTree(data.split("::")[1]);
+						// make gets more robuts.
+						platformId = node.get("platformId").asText();
+						instanceId = node.get("instanceId").asText();
+					} catch (IOException e) {
+						// TODO Auto-generated catch block
+						e.printStackTrace();
+					}
+				}
+			}
+			
+		}
+		
+		PerformanceData data = new PerformanceData(this.functionName,
+				instanceId,
+				platformId,
+				startTime,
+				endTime,
+				-1.0,
+				preciseDuration,
+				-1,
+				-1,
+				-1
+				);
+		return data;
+	}
+
+	/**
+	 * Creates a map to group the log entries returned by the Stackdriver API.
+	 * The key is the platform's execution_id.
+	 */
+	private Map<String, List<LogEntry>> generateEventMapByExecutionId(List<LogEntry> logEntries) {
+		Map<String, List<LogEntry>> eventMap = new HashMap<>();
+
+		for (LogEntry logEntry : logEntries) {
+			String executionId = logEntry.getLabels().get("execution_id");
+			if (eventMap.containsKey(executionId)) {
```

Comment:
This could be simplified by first checking if the executionId exists. If not create and add the list. Afterwards add the log entry:

```
if (!eventMap.containsKey(executionId)) {
    eventMap.put(executionId, new ArrayList<LogEntry>());
}

eventMap.get(executionId).add(logEntry);
```